contracts/BaseWrapper.sol

```diff
@@ -0,0 +1,162 @@
+                .mul(best.pricePerShare()) // Adjust by price of best
+                .div(10**best.decimals())
+                .div(10**3); // Add 3 extra decimals of precision
+    }
+
+    function _withdraw(
+        address sender,
+        address receiver,
+        uint256 amount,
+        bool withdrawFromBest // If true, also withdraw from `best`
+    ) internal returns (uint256 withdrawn) {
+        VaultAPI best = bestVault();
+
+        VaultAPI[] memory vaults = allVaults();
+        _updateVaultCache(vaults);
+
+        for (uint256 id = 0; id < vaults.length; id++) {
+            if (!withdrawFromBest && vaults[id] == best) {
+                continue; // Don't withdraw from the best
+            }
+
+            uint256 shares = vaults[id].balanceOf(sender);
+
+            // NOTE: No need for allowance check if we are withdrawing the balance from this contract
+            if (shares > 0 && (sender == address(this) || vaults[id].allowance(sender, address(this)) >= shares)) {
+                // NOTE: There is a maximum withdrawal size from each vault (in shares)
+                uint256 maxWithdrawal = vaults[id].maxAvailableShares();
+
+                if (maxWithdrawal < shares) shares = maxWithdrawal;
+
+                // NOTE: use `uint256(-1)` for withdrawing everything
+                uint256 withdrawalLeft = amount
+                    .sub(withdrawn) // NOTE: Changes every iteration
+                    .mul(vaults[id].pricePerShare()) // NOTE: Every Vault is different
+                    .div(10**vaults[id].decimals());
+
+                if (withdrawalLeft == 0) break; // withdrawn as much as we needed
+                if (withdrawalLeft < shares) shares = withdrawalLeft;
+
+                // NOTE: No need for share transfer if we are migrating the balance of this contract
+                if (sender != address(this)) vaults[id].transferFrom(sender, address(this), shares);
+                withdrawn = withdrawn.add(vaults[id].withdraw(shares, receiver));
+            }
+        }
+        // `receiver` now has `withdrawn` tokens as balance
+    }
+
+    function _migrate(address account) internal returns (uint256 migrated) {
+        VaultAPI best = bestVault();
+
+        uint256 alreadyDeposited = best.balanceOf(account).mul(best.pricePerShare()).div(10**best.decimals());
+        uint256 amountToMigrate = totalBalance(account).sub(alreadyDeposited);
+
+        uint256 depositLeft = best.depositLimit().sub(best.totalAssets());
+        if (amountToMigrate > depositLeft) amountToMigrate = depositLeft;
+
+        if (amountToMigrate > 0) {
+            migrated = _withdraw(account, address(this), amountToMigrate, false); // `false` = don't withdraw from `best`
+            require(migrated == amountToMigrate);
+            require(migrated == _deposit(address(this), account, migrated, false)); // `false` = don't do `transferFrom` because it's already local
```

Comment:
after all the side effects on migrate should the invariant be true:

`totalAssetsBeforeMigrate >=totalAssetsAfterMigrate` ?

Reply:
Yeah, it should be, since the prices of the relevant deposited shares should equal out before and after.

NOTE: I had to update this check to do an abs value check of `< 10` because it caused a failure sometimes. This should be appropriate in most circumstances, but something to be aware of. Also note that the value returned by `_deposit` is approximate